### PR TITLE
Lazy evaluation the ring projects

### LINKED ISSUE #1626: pkglistgen fails to work on Leap 15.1 after expanded_repos() moved stagingapi
With the all scopes, pkglistgen fails to work on Leap 15.1 when through `:ARM` due to can not find matched config.

    Jul 31 11:10:59 repochecker osrt-pkglistgen[14247]:   File "/usr/bin/osrt-pkglistgen", line 1152, in do_update_and_solve
    Jul 31 11:10:59 repochecker osrt-pkglistgen[14247]:     self.do_update_and_solve(subcmd, copy.deepcopy(opts))
    Jul 31 11:10:59 repochecker osrt-pkglistgen[14247]:   File "/usr/bin/osrt-pkglistgen", line 1182, in do_update_and_solve
    Jul 31 11:10:59 repochecker osrt-pkglistgen[14247]:     self.repos = self.tool.expand_repos(opts.project, main_repo)
    Jul 31 11:10:59 repochecker osrt-pkglistgen[14247]:   File "/usr/bin/osrt-pkglistgen", line 528, in expand_repos
    Jul 31 11:10:59 repochecker osrt-pkglistgen[14247]:     return StagingAPI(self.apiurl, project).expanded_repos(repo)
    Jul 31 11:10:59 repochecker osrt-pkglistgen[14247]:   File "/usr/share/openSUSE-release-tools/osclib/stagingapi.py", line 85, in __init__
    Jul 31 11:10:59 repochecker osrt-pkglistgen[14247]:     if self.crings:
    Jul 31 11:10:59 repochecker osrt-pkglistgen[14247]:   File "/usr/share/openSUSE-release-tools/osclib/stagingapi.py", line 102, in __getattr__
    Jul 31 11:10:59 repochecker osrt-pkglistgen[14247]:     value = conf.config[self.project][key]
    Jul 31 11:10:59 repochecker osrt-pkglistgen[14247]: KeyError: 'openSUSE:Leap:15.1:ARM'

### PR DESCRIPTION
**Pull request by coolo**
So we can use the StagingAPI also for projects below the target
project

Fixes #1626

### CODE REVIEW

**Comment by coolo** [comment]
Possibly StagingAPI should be split? It feels like an abuse to me

File: osclib/stagingapi.py [hunk 1]
```diff
@@ -110,6 +101,20 @@ def __getattr__(self, attr):
         # Raise AttributeError like normal.
         return self.__getattribute__(attr)
 
+    # lazy load
+    def rings(self):
```
Comment (aplanas) [inline osclib/stagingapi.py]: Uhm no, this is a bug as Python is not a Lisp 2 -a-like language:

```
>>> class A:
...   def __init__(self):
...     self.a = 0
...   def a(self):
...     if self.a:
...        print 'no zero'
... 
>>> kk = A()
>>> kk.a
0
>>> kk.a()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'int' object is not callable
```

Rename it to _rings(), or get_rings()

Comment (lnussel) [reply osclib/stagingapi.py]: or use \@property ?

Comment (coolo) [reply osclib/stagingapi.py]: ok, I'm out

Comment (lnussel) [reply osclib/stagingapi.py]: https://github.com/openSUSE/openSUSE-release-tools/blob/master/ReviewBot.py#L152

File: osclib/stagingapi.py [hunk 2]
```diff
@@ -110,6 +102,20 @@ def __getattr__(self, attr):
...
+
+        # If the project support rings, inititialize some variables.
+        if self.crings:
+            self._rings = (
+                '{}:0-Bootstrap'.format(self.crings),
+                '{}:1-MinimalX'.format(self.crings)
+            )
+        else:
+            self._rings = []
+
```
Comment (nilxam) [inline osclib/stagingapi.py]: initialing `rings` is no longer happened in `__init__`, I thought first time calling `self.rings` has nothing returned but second time does?

Comment (lnussel) [reply osclib/stagingapi.py]: so should be

```
if self._rings is None:
  initialize it
  
return self._rings
```

Comment (coolo) [reply osclib/stagingapi.py]: Did I mention it's just too warm ? :(

**Comment by coolo** [comment]
So review round 4 then

**Pull request merged**